PSPs/psp-22.md

```diff
@@ -0,0 +1,210 @@
+- **Status:** Draft
+- **Created:** 2021-06-19
+- **Reference Implementation:** [OpenBrush](https://github.com/Supercolony-net/openbrush-contracts/blob/main/contracts/token/psp20/traits.rs)
+
+
+## Summary
+
+A standard interface for Ink! tokens.
+
+This proposal aims to define the standard token in ink! smart contracts, just like [EIP-20](https://github.com/ethereum/EIPs/edit/master/EIPS/eip-20.md) for Ethereum ecosystem.
+
+## Importance
+Currently, while there is no standard, every contract will have different signature. Thus, no interoperability is possible. This proposal aims to resolve that by having one **trait**(interface) that shares the
+same **trait naming** between all implementations, as naming of trait affects the identifiers of functions in this trait.
+
+## Implementation
+
+Example implementation:
+
+- [OpenBrush](https://github.com/Supercolony-net/openbrush-contracts/blob/main/contracts/token/psp20/traits.rs)
+
+## Motivation
+
+A standard interface allows any Ink! tokens on Polkadot/Kusama to be re-used by other applications: from wallets to decentralized exchanges.
+
+
+## Motivation for having a standard separate from ERC20
+
+Due to different nature of ink!, the standard should have ink!-specific rules and methods,
+therefore PSP-22 differs from ERC-20 in its implementation.
+
+Also, this standard proposal defines an extensive method list in the trait (interface). Unlike ERC20, it includes `increase_allowance` & `decrease_allowance`, and defines metadata fields as part of a separate trait.
+Another difference is that it has `PSP22Receiver` trait, and `on_received` method is called at the end of transfer if the recipient is a contract.
+
+
+## Specification
+1. [Traits](#Traits)
+2. [Types](#Types)
+3. [Events](#Events)
+4. [Errors](#Errors)
+
+### Traits
+
+```rust
+/// PSP22 is a trait of Fungible Token Standard.
+pub trait PSP22 {
+ /// Returns the total token supply.
+ fn total_supply(&self) -> Balance;
+
+ /// Returns the account balance for the specified `owner`.
+ /// Returns `0` if the account is non-existent.
+ fn balance_of(&self, owner: AccountId) -> Balance;
+
+ /// Returns the amount which `spender` is still allowed to withdraw from `owner`.
+ /// Returns `0` if no allowance has been set.
+ fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance;
+
+ /// Transfers `value` amount of tokens from the caller's account to account `to`
+ /// with additional `data` with no specified format.
+ /// Emits a `Transfer` event on success.
+ /// 
+ /// This method also calls [on_received](#on_received) method on `to`.
+ /// 
+ /// **Errors**
+ /// * Panics with `InsufficientBalance` error if there are not enough tokens on
+ /// the caller's account balance.
+ /// * Panics with `ZeroSenderAddress` error if sender's address is zero.
+ /// * Panics with `ZeroRecipientAddress` error if recipient's address is zero.
```

Comment:
@athei Does the `contracts` pallet currently have the ability of forwarding/exposing the `panic` payload or is it just the `ContractTrapped` without further information?

Reply:
Panic is a Rust concept. In ink! a panic translates to emitting an `unreachable` instruction and nothing more. We even build `core` with `immediate-abort` so that no string handling stuff is pulled in.

In order to bubble up error information on failure a contract can use `seal_return` with the `REVERT` flag set. The `data` can then be used to encode the error. Not sure if we need another mechanism. `seal_return` stops execution immediately just like a panic would. One could even change the ink! panic handler to call this function instead of emitting `unreachable`.

That said, the marked specification is not well defined. It should be defined in terms of supervisor functions like. For example: Returns with `REVERT` flag set and `ZeroSenderAddress` encoded in the return buffer.

Reply:
> Panic is a Rust concept. In ink! a panic translates to emitting an `unreachable` instruction and nothing more. We even build `core` with `immediate-abort` so that no string handling stuff is pulled in.
> 
> In order to bubble up error information on failure a contract can use `seal_return` with the `REVERT` flag set. Error information can be passed in `data`. Not sure if we need another mechanism. `seal_return` stops execution immediately just like a panic would. One could even change the ink! panic handler to call this function instead of emitting `unreachable`.
> 
> That said, the marked specification is not well defined. It should be defined in terms of supervisor functions like. For example: Returns with `REVERT` flag set and `ZeroSenderAddress` encoded in the return buffer.

I think the ink! team must decide how better to force revert of the transaction(API). The main idea was that the contract will force a revert of the transaction instead of returning an error. You are right, we will change the description to use "revert" instead of "panic". 

BTW, when we are using `seal_return`, the substrate node marks this execution as `Success`. But I think it must be marked as `Reverted` because we set `REVERT` flag.

Reply:
> BTW, when we are using seal_return, the substrate node marks this execution as Success. But I think it must be marked as Reverted because we set REVERT flag.

The calling contract decides how to handle a reverting sub call. However, for the top level call you are right. It should be treated as an error but it isn't: https://github.com/paritytech/substrate/issues/9338